### PR TITLE
MergeSamFiles accept SO:UNKNOWN

### PR DESCRIPTION
**Pull request by a-filipp0v**
### Description

Request is associated with Picard's issue https://github.com/broadinstitute/picard/issues/796

As said in that issue MergeSamFiles can't accept SAM-file with SO:UNKNOWN. That behavior is expected, because tag's value is in improper case(in accordance with SAM-format specification, SO value must be in lowercase). We've found that wrong tag case also provides an exception.

The idea of implementation is that htsjdk must check case of header's tags and values. If tag/value itself is valid, but in improper case(e.g. lowercase instead of uppercase) - it should be converted to proper case before sending the result to output(without changing the source file). Also, there must be a message if that conversion was made.

### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

**Comment by yfarjoun** [comment]
Personally I'm opposed to this change. 

I do not see why htsjdk needs to be made more complicated in order to accept non-SAM-spec compliant input. Its hard enough as it is even with SAM-compliant input and this will make the code more complicated. What if someone has a typo in the header and they spelled the sort order "unkwnon"? Or perhaps they use a different "locale" and the used "SO: desconocido" ? should htsjdk also accept that? 

That off my chest, I'm not a maintainer...so I'd love to hear from them before even looking at the details of the implementation: @lbergelson @tfenne  @jacarey

**Comment by yfarjoun** [comment]
I think that the correct solution is to fix the offending software so that it produces compliant BAM files and to write a simple script that fixes "broken" headers.

**Comment by yfarjoun** [comment]
Alternatively a better  (in my view) solution would be an ability to ignore (when validation=silent or leniant) an unknown sort order and use "unknown" when that happens (instead of lowercasing)

**Comment by magicDGS** [comment]
I also prefer that HTSJDK is SAM-compliant and do not allow non-supported `SO`...

**Comment by jacarey** [comment]
I agree with @yfarjoun

**Comment by yfarjoun** [comment]
@a-filipp0v could you modify the code as follows:
1. remove the lower-casing and the logs related to that
2. if SortOrder.valueOf(unmodified) throws an error, AND VALIDATION_STRINGENCY!=STRICT, use SortOrder.UNKNOWN and either emit a warning or not depending on VALIDATION_STRINGENCY.

I think that this will both solve the issue at hand and not introduce un-needed complexity.

As a bonus, I think that your tests should still be useful!

**Comment by a-filipp0v** [comment]
Thank you for your comments. Hope we've understood you correctly in case of non-conforming SO tag values.
Our implementation has been reworked according to your requests. Also, we've reworked previous tests in SamFileHeaderTest, added few more and removed one test case from MergeSamFilesTest because it's redundant after our implementation.

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 1]
```diff
@@ -324,12 +324,18 @@ public void setAttribute(final String key, final Object value) {
...
     @Override
     public void setAttribute(final String key, final String value) {
+        String tempVal = value;
         if (key.equals(SORT_ORDER_TAG)) {
-            this.sortOrder = null;
+            try {
+                this.sortOrder = null;
+                tempVal = SortOrder.valueOf(value).toString();
+            } catch (IllegalArgumentException e) {
+                tempVal = "unknown";
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: use `SortOrder.unknown.toString()` here.

File: src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java [hunk 2]
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
...
                     continue;
                 }
+                validateSortOrderValue(keyAndValue);
                 mKeyValuePairs.put(keyAndValue[0], keyAndValue[1]);
             }
             lineValid = true;
         }
 
+        private void validateSortOrderValue(String[] value) {
+            if ("SO".equals(value[0])) {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java]: use SAMFileHeader.SORT_ORDER_TAG

File: src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java [hunk 3]
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
...
+                } catch (IllegalArgumentException e) {
+                    if (validationStringency == ValidationStringency.STRICT) {
+                        throw new SAMFormatException("Found non conforming header SO tag: "
+                                                     + value[1]
+                                                     + ", exiting because VALIDATION_STRINGENCY=STRICT");
+                    } else if (validationStringency == ValidationStringency.LENIENT) {
+                        log.warn("Found non conforming header SO tag: "
+                                 + value[1] + ". Treating as 'unknown'.");
+                    }
+                    value[1] = "unknown";
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java]: use SortOrder.unknown.toString()

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 4]
```diff
@@ -324,12 +324,18 @@ public void setAttribute(final String key, final Object value) {
      */
     @Override
     public void setAttribute(final String key, final String value) {
+        String tempVal = value;
         if (key.equals(SORT_ORDER_TAG)) {
-            this.sortOrder = null;
+            try {
+                this.sortOrder = null;
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: Why move this line into the try clause?

File: src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java [hunk 5]
```diff
@@ -69,6 +70,8 @@
     private static final Pattern FIELD_SEPARATOR_RE = Pattern.compile(FIELD_SEPARATOR);
 
     public static final String COMMENT_PREFIX = HEADER_LINE_START + HeaderRecordType.CO.name() + FIELD_SEPARATOR;
+    private static final Log log = Log.getInstance(SAMTextHeaderCodec.class);
+
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java]: remove extra newline

File: src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java [hunk 6]
```diff
@@ -233,8 +236,8 @@ private void parseHDLine(final ParsedHeaderLine parsedHeaderLine) {
         try {
             if (soString != null) SAMFileHeader.SortOrder.valueOf(soString);
         } catch (IllegalArgumentException e) {
-            reportErrorParsingLine(HEADER_LINE_START + parsedHeaderLine.getHeaderRecordType() +
-                            " line has non-conforming SO tag value: "+ soString + ".",
+                reportErrorParsingLine(HEADER_LINE_START + parsedHeaderLine.getHeaderRecordType() +
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java]: revert extra tab

File: src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java [hunk 7]
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
                             SAMValidationError.Type.HEADER_TAG_MULTIPLY_DEFINED, null);
                     continue;
                 }
+                validateSortOrderValue(keyAndValue);
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java]: why is this needed here? could you put it instead in SamFileValidator? perhaps in the ValidateHeader method?

note that the validation methods create a tally of validation failures... and this should probably do the same.

File: src/test/java/htsjdk/samtools/ValidateSamFileTest.java [hunk 8]
```diff
@@ -552,12 +552,6 @@ public void duplicateReadsOutOfOrder() throws Exception {
                         "@RG\tID:0\tSM:Hi,Mom!\n" +
                         "E\t147\tchr1\t15\t255\t10M\t=\t2\t-30\tCAACAGAAGC\t)'.*.+2,))\tU2:Z:CAA";
 
-        final String SOTagCorrectlyProcessTestData =
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/ValidateSamFileTest.java]: why is this removed?

**Comment by a-filipp0v** [comment]
Hello, @yfarjoun , thanks for your reply.

According to your previous comments, we've tried to move an appropriate check to SamFileValidator, but found that it still might make sense to leave our checks in SAMTextHeaderCodec and SAMFileHeader.

First, the check in SamFileValidator will be called only when the ValidateSamFile is used, which does not prevent an error when using any tools separately (for example, MergeSamFile as in this issue).

Secondly, SamFileValidator does not contain separate data about Validation Stringency, because this parameter is parsed from Picard during start.

We suggest leaving checks in SAMTextHeaderCodec and SAMFileHeader AND adding a similar one in SamFileValidator. What do you think about it?

**Comment by lbergelson** [comment]
@yfarjoun Since you did the initial review, could you take a look at this again when you get a chance?

**Comment by yfarjoun** [comment]
@a-filipp0v thanks for the explanation. I've removed my comment relating to the validation.

**Comment by yfarjoun** [comment]
@a-filipp0v sorry for the delay. please respond to the review comments.

**Comment by yfarjoun** [comment]
@a-filipp0v just so we know, are you planning to get back to this PR?

File: src/test/java/htsjdk/samtools/ValidateSamFileTest.java (hunk 8 above)
Comment (IanaKabakova) [reply src/test/java/htsjdk/samtools/ValidateSamFileTest.java]: thank you for your reply. We've removed this test case since as @a-filipp0v said, it's redundant after our implementation.

**Comment by IanaKabakova** [comment]
@yfarjoun sorry for the delay and thank you for your reply. I made some fixes according to your notes. Could you take a look at PR again?

File: src/test/java/htsjdk/samtools/ValidateSamFileTest.java (hunk 8 above)
Comment (yfarjoun) [reply src/test/java/htsjdk/samtools/ValidateSamFileTest.java]: not sure I understand...if it's just "Redundant" but not exactly equals, I'd ask that it be left in.

**Comment by yfarjoun** [comment]
Thanks for the changes. Please leave tests as is, unless they no longer pass.

File: src/test/java/htsjdk/samtools/ValidateSamFileTest.java (hunk 8 above)
Comment (IanaKabakova) [reply src/test/java/htsjdk/samtools/ValidateSamFileTest.java]: This case does not pass now, since we no longer treat SO:NOTKNOWN as error as test expects.

Comment (yfarjoun) [reply src/test/java/htsjdk/samtools/ValidateSamFileTest.java]: got it

**Comment by yfarjoun** [comment]
@jacarey @tfenne @lbergelson any opinions about this small API change?

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 9]
```diff
@@ -324,12 +324,18 @@ public void setAttribute(final String key, final Object value) {
      */
     @Override
     public void setAttribute(final String key, final String value) {
+        String tempVal = value;
         if (key.equals(SORT_ORDER_TAG)) {
             this.sortOrder = null;
+            try {
+                tempVal = SortOrder.valueOf(value).toString();
```
Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: In `setSortOrder()` the code uses `.name()` instead of `.toString()` which I think is clearer; either way the code should be consistent.

Also, it seems odd that `setSortOrder()` sets `sortOrder` to non-null, if the argument is a valid `SortOrder`, but this method sets it to `null`. Maybe this method should just call `setSortOrder()` when `key` is `SORT_ORDER_TAG`?

File: src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java [hunk 10]
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
...
                 mKeyValuePairs.put(keyAndValue[0], keyAndValue[1]);
             }
             lineValid = true;
         }
 
+        private void validateSortOrderValue(String[] value) {
```
Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java]: I think it would be slightly clearer if the caller extracted `value[0]` and `value[1]` so they could be declared as `String key` and `String value` here.

File: src/main/java/htsjdk/samtools/SAMFileHeader.java (hunk 9 above)
Comment (yfarjoun) [reply src/main/java/htsjdk/samtools/SAMFileHeader.java]: @pshapiro4broad that's how the old implementation also does it...it let's the super-class deal with it...take a look at htsjdk.samtools.SAMFileHeader#getSortOrder, I think that might clarify things up (and note that this.sortOrder is private)

**Comment by yfarjoun** [comment]
No response from the maintainers for a while now, I suspect that no-one objects....if I'm wrong, please revert and accept my apologies.
